### ISSUE TITLE
we need Embeddings Lookup this method

### ISSUE THREAD

**Issue by mullerhai** [tensorflow/java#416]
Hi :
    I wanna to invoke tf.nn.embedding_lookup to make embedding layer ,but the java version has not these part code.please  make it real work in java ? thank you ,

**Comment by saudet** [comment]
FYI, it looks like the C++ API of PyTorch implements this functionality, which we can easily use from Java:
https://github.com/bytedeco/javacpp-presets/blob/master/pytorch/src/gen/java/org/bytedeco/pytorch/EmbeddingImpl.java

**Comment by karllessard** [comment]
Hi @mullerhai , did you tried to use [`tf.gather`](https://github.com/tensorflow/java/blob/5210c7e7aff5d0fb615ffb5a40769c632a37db3a/tensorflow-core/tensorflow-core-api/src/gen/annotations/org/tensorflow/op/Ops.java#L2640) instead? According to TF [documentation](https://www.tensorflow.org/api_docs/python/tf/nn/embedding_lookup), `embedding_lookup` is just a generalization of `gather`.

**Comment by saudet** [comment]
If you could port that to Java, I'm sure we will want to merge this, yes.

**Comment by karllessard** [comment]
Thanks @mullerhai ,

Again, the "happy path" of `embedding_lookup` (getting raw embeddings from a single tensor) can be replaced by a single call to `tf.gather`. But if you could write/port a wrapper that does also the fancy stuff (partitioning, norm, transform, ...), that would be certainly a great addition to our Java framework.

**Comment by mullerhai** [comment]
> Thanks @mullerhai ,
> 
> Again, the "happy path" of `embedding_lookup` (getting raw embeddings from a single tensor) can be replaced by a single call to `tf.gather`. But if you could write/port a wrapper that does also the fancy stuff (partitioning, norm, transform, ...), that would be certainly a great addition to our Java framework.

yeah, if the tf java framework has   embedding look up method  then  we will have embedding layer, it will benefit for all users

**Comment by karllessard** [comment]
Talking of embedding layers… @JimClarke5 , is this something you have worked on as part of the first phase for layers?

**Comment by JimClarke5** [comment]
Exactly what do you mean by embedding layers? The Sequential model can contain a network of layers if that is what you mean.

I have not worked on layers for the simpler models, the plan is to get the  basic model plumbing accepted, then add more layers.

Also, I have not tackled saving the model in a SavedModelBundle yet. 

There is a lot of work yet for models and layers.

Jim

**Comment by karllessard** [comment]
Yes I was asking if you’ve already started to port the [Keras embedding layer](https://keras.io/api/layers/core_layers/embedding/), to make sure we are not duplicating the efforts here.

**Comment by JimClarke5** [comment]
I haven’t started it. There is some plumbing for Layers you will need, like the Layer Interface and BaseLayer abstract class.

Jim
